src/editable_tree/mod.rs

```diff
@@ -190,56 +322,113 @@ impl<'arena, Node: Ast<'arena>> DAG<'arena, Node> {
         self.history_index = self.root_history.len() - 1;
     }
 
-    /// Updates the internal state so that the tree now contains `new_node` in the position of the
-    /// `cursor`.
-    pub fn replace_cursor(&mut self, new_node: Node) {
+    /// Replaces the current cursor with a node represented by `c`
+    pub fn replace_cursor(&mut self, c: char) -> EditResult {
+        /* CHECK THE VALIDITY OF ARGUMENTS */
+
+        // Cache the node under the cursor, since finding the cursor involves non-trivial amounts
+        // of work
+        let cursor = self.cursor();
+        // Short circuit if the char to insert couldn't correspond to a valid child
+        if !cursor.is_replace_char(c) {
+            return EditResult::Err(EditErr::CannotReplace(c));
+        }
+
+        /* PERFORM THE ACTION */
+
+        // We require that `c` corresponds to a valid node (because `is_replace_char` returned
+        // `true`), so we can unwrap
+        let new_node = cursor.from_char(c).unwrap();
```

Comment:
Why not this?

```rust
let new_node = cursor.from_char(c).ok_or(EditErr::CannotReplace(c))?;
```

And get rid of the `if` above.

Reply:
This won't remove the `if` without changing the effect of the code.  `Ast::from_char` will always return a new node even if it would make the tree invalid (I'd quite like to remove the `&self` parameter to make this explicit...) - whereas `Ast::is_replace_char` checks that a node represented by a given char is valid in the given location.

However, this is still  a good spot - it'd be better to return an error than use `unwrap`.

Reply:
Cool - now fixed in all locations